src/cpp/dynamic/DiscoveryDatabase.cpp

```diff
@@ -26,43 +26,84 @@ namespace ddsrouter {
 bool DiscoveryDatabase::topic_exists(
         const RealTopic& topic) const
 {
-    // TODO
-    throw UnsupportedException("DiscoveryDatabase::topic_exists is not implemented yet");
+    std::shared_lock<std::shared_timed_mutex> lock(mutex_);
+
+    // Loop over every Endpoint checking if topic matches
+    // It is not required to be reference
+    for (auto entity : entities_)
+    {
+        if (entity.second.topic() == topic)
+        {
+            return true;
+        }
+    }
+    return false;
 }
 
 bool DiscoveryDatabase::endpoint_exists(
         const Guid& guid) const
 {
-    // TODO
-    throw UnsupportedException("DiscoveryDatabase::endpoint_exists is not implemented yet");
+    std::shared_lock<std::shared_timed_mutex> lock(mutex_);
+    return entities_.find(guid) != entities_.end();
 }
 
 ReturnCode DiscoveryDatabase::add_or_modify_endpoint(
         const Endpoint& new_endpoint)
 {
-    // TODO
-    throw UnsupportedException("DiscoveryDatabase::add_or_modify_endpoint is not implemented yet");
+    std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
+    auto it = entities_.find(new_endpoint.guid());
+    if (it != entities_.end())
+    {
+        // Already exists, modify it
+        it->second = new_endpoint;
+        return ReturnCode::RETCODE_OK;
+    }
+    else
+    {
+        // Add it to the dictionary
+        entities_.insert(std::pair<Guid, Endpoint>(new_endpoint.guid(), new_endpoint));
+        return ReturnCode::RETCODE_OK;
+    }
 }
 
 ReturnCode DiscoveryDatabase::erase_endpoint(
         const Guid& guid_of_endpoint_to_erase)
 {
-    // TODO
-    throw UnsupportedException("DiscoveryDatabase::erase_endpoint is not implemented yet");
+    std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
+    auto erased = entities_.erase(guid_of_endpoint_to_erase);
+
+    if (erased == 0)
+    {
+        return ReturnCode::RETCODE_NO_DATA;
+    }
+    else
+    {
+        return ReturnCode::RETCODE_OK;
+    }
 }
 
 ReturnCode DiscoveryDatabase::erase_endpoint(
         const Endpoint& endpoint_to_erase)
 {
-    // TODO
-    throw UnsupportedException("DiscoveryDatabase::erase_endpoint is not implemented yet");
+    // Mutex is taken by erase_endpoint with guid, so it is not needed here
+    return erase_endpoint(endpoint_to_erase.guid());
 }
 
 Endpoint DiscoveryDatabase::get_endpoint(
         const Guid& endpoint_guid) const
 {
-    // TODO
-    throw UnsupportedException("DiscoveryDatabase::get_endpoint is not implemented yet");
+    std::shared_lock<std::shared_timed_mutex> lock(mutex_);
+
+    auto it = entities_.find(endpoint_guid);
+    if (it == entities_.end())
+    {
+        // TODO: Add log warning
+        return Endpoint();
```

Comment:
That was the first idea. But I wanted to avoid two things: 
1. Allowing to create an empty Endpoint 
2. Use method arguments as return variables.
And if the endpoint is returned as argument, then the object that asks must create an empty Endpoint, so there is no way to avoid 1.; thus I decided to do it this way as it avoids 2.

NIT: Using exceptions would avoid both, but I preferred this approach.